hibernate-core/src/main/java/org/hibernate/sql/results/graph/entity/AbstractEntityInitializer.java

```diff
@@ -117,10 +121,15 @@ protected AbstractEntityInitializer(
 			this.identifierAssembler = identifierResult.createResultAssembler(
 					new AssemblerCreationState() {
 						@Override
-						public Initializer resolveInitializer(NavigablePath navigablePath, Supplier<Initializer> producer) {
+						public Initializer resolveInitializer(
+								NavigablePath navigablePath,
+								ModelPart fetchedModelPart,
+								Supplier<Initializer> producer) {
 							for ( int i = 0; i < identifierInitializers.size(); i++ ) {
 								final Initializer existing = identifierInitializers.get( i );
-								if ( existing.getNavigablePath().equals( navigablePath ) ) {
+								if ( existing.getNavigablePath().equals( navigablePath )
+										&& fetchedModelPart.getNavigableRole().equals(
```

Comment:
Just noticed some format inconsistency. Sometimes we split BEFORE boolean operator, sometimes we split AFTER it.